realm-jni/build.gradle

```diff
@@ -89,18 +104,75 @@ task checkProperties(group: 'check', description: 'Check the user provided gradl
     }
 }
 
-// TODO: Split this task into smaller subtasks
+ext.coreArchiveFile = project.file("../core-android-${project.coreVersion}.tar.gz")
+ext.coreDir = new File(buildDir, "core-${project.coreVersion}")
+
+def coreDownloaded = false
+
 task downloadCore(group: 'build setup', description: 'Download the latest version of realm core', dependsOn: checkProperties) {
-    outputs.dir "../core-${project.coreVersion}"
+    outputs.upToDateWhen {
+        if (project.forceDownloadCore || !project.coreArchiveFile.exists()) {
+            return false
+        }
+        if (!isHashCheckingEnabled()) {
+            println 'skipped hash checking(empty \'coreSha256Hash\').'
+            return true
+        }
+
+        MessageDigest sha = MessageDigest.getInstance("SHA-256");
+        Formatter hexHash = new Formatter()
+        sha.digest(project.coreArchiveFile.bytes).each { b -> hexHash.format('%02x', b) }
+        return project.coreSha256Hash.equals(hexHash.toString())
+    }
+
     doLast {
+        if (project.forceDownloadCore) {
+            println 'forceDownloadCore is true.'
+        }
         download {
             src "http://static.realm.io/downloads/core/realm-core-android-${project.coreVersion}.tar.gz"
-            dest new File(buildDir, "core-android-${project.coreVersion}.tar.gz")
-            onlyIfNewer true
+            dest project.coreArchiveFile
+            onlyIfNewer false
+        }
+        coreDownloaded = true
+        if (!isHashCheckingEnabled()) {
+            println 'skipped hash checking(empty \'coreSha256Hash\').'
+        }
+    }
+}
+
+task downloadAndVerifyCore(type: de.undercouch.gradle.tasks.download.Verify, dependsOn: downloadCore) {
+    src       project.coreArchiveFile
+    algorithm 'SHA-256'
+    checksum  isHashCheckingEnabled() ? project.coreSha256Hash : ''
+}
+
+task deployCore(group: 'build setup', description: 'Deploy the latest version of realm core') {
+    if (isHashCheckingEnabled()) {
```

Comment:
Minor thing: It would feel more encapsulated if this if statement was part of just a `downloadCore` task, i.e. merge `downloadCore` and `downloadAndVerifyCore` instead of having having the hash being referenced across 3 functions.